ruby/ext/google/protobuf_c/storage.c

```diff
@@ -178,9 +178,42 @@ void native_slot_set_value_and_case(const char* name,
       if (CLASS_OF(value) == CLASS_OF(Qnil)) {
         value = Qnil;
       } else if (CLASS_OF(value) != type_class) {
-        rb_raise(cTypeError,
-                 "Invalid type %s to assign to submessage field '%s'.",
-                rb_class2name(CLASS_OF(value)), name);
+        // check for possible implicit conversions
+        VALUE converted_value = NULL;
+        char* field_type_name = rb_class2name(type_class);
+        char* value_type_name = rb_class2name(CLASS_OF(value));
+
+        // Time -> Google::Protobuf::Timestamp
+        if (strcmp(field_type_name, "Google::Protobuf::Timestamp") == 0 &&
+            strcmp(value_type_name, "Time") == 0) {
+          VALUE hash = rb_hash_new();
+          rb_hash_aset(hash, rb_str_new2("seconds"), rb_funcall(value, rb_intern("to_i"), 0));
+          rb_hash_aset(hash, rb_str_new2("nanos"), rb_funcall(value, rb_intern("nsec"), 0));
+          VALUE args[1] = { hash };
+          converted_value = rb_class_new_instance(1, args, type_class);
+        }
+
+        // Number -> Google::Protobuf::Duration
+        if (strcmp(field_type_name, "Google::Protobuf::Duration") == 0 &&
+            (RB_TYPE_P(value, T_FIXNUM) || RB_TYPE_P(value, T_FLOAT))) {
```

Comment:
I would still rather see a check for `Numeric` rather than `Fixnum` or `Float`. Users should be allowed to pass in any `Numeric` objects, such as `BigDecimal` or `Rational`.

Reply:
Does [`num_exact`](https://github.com/ruby/ruby/blob/f284eb3d54df6d1bb754986f9f5272136ff73466/time.c#L492-L538) do what we would need done here?

Reply:
I don't think `num_exact` is the right method (and we only need to check the type not get the value). I could probably just enumerate all the `Numeric` types?

Reply:
Does `rb_obj_is_kind_of(value, rb_cNumeric)` do this?

Reply:
Ah, I think you've found it! I've replaced it with `rb_obj_is_kind_of`